locator/network_topology_strategy.cc

```diff
@@ -302,40 +302,55 @@ effective_replication_map_ptr network_topology_strategy::make_replication_map(ta
-// are covered with at least one tablet. In other words, the value is
+// are covered with at least `min_per_shard_tablet_count` tablets. In other words, the value is
 //
 //    initial_tablets = max(nr_shards_in(dc) / RF_in(dc) for dc in datacenters)
 //
 
-static unsigned calculate_initial_tablets_from_topology(const schema& s, token_metadata_ptr tm, const std::unordered_map<sstring, size_t>& rf) {
+static unsigned calculate_initial_tablets_from_topology(const schema& s, token_metadata_ptr tm, const std::unordered_map<sstring, size_t>& rf, double min_per_shard_tablet_count = 0) {
     unsigned initial_tablets = std::numeric_limits<unsigned>::min();
-    for (const auto& dc : tm->get_datacenter_token_owners()) {
-        unsigned shards_in_dc = 0;
-        unsigned rf_in_dc = 1;
-
-        for (const auto& ep : dc.second) {
-            const auto* node = tm->get_topology().find_node(ep);
-            if (node != nullptr) {
-                shards_in_dc += node->get_shard_count();
-            }
+    std::unordered_map<sstring, unsigned> shards_per_dc_map;
+    tm->get_topology().for_each_node([&] (const node& n) {
+        if (tm->is_normal_token_owner(n.host_id())) {
+            shards_per_dc_map[n.dc_rack().dc] += n.get_shard_count();
         }
-
-        if (auto it = rf.find(dc.first); it != rf.end()) {
-            rf_in_dc = it->second;
+    });
+    for (const auto& [dc, rf_in_dc] : rf) {
+        if (!rf_in_dc) {
+            continue;
+        }
+        unsigned shards_in_dc = shards_per_dc_map[dc];
+        unsigned tablets_in_dc = (shards_in_dc + rf_in_dc - 1) / rf_in_dc;
+        if (min_per_shard_tablet_count) {
+            auto min_tablets_in_dc = std::ceil((double)(min_per_shard_tablet_count * shards_in_dc) / rf_in_dc);
+            tablets_in_dc = std::max<unsigned>(tablets_in_dc, min_tablets_in_dc);
         }
-
-        unsigned tablets_in_dc = rf_in_dc > 0 ? (shards_in_dc + rf_in_dc - 1) / rf_in_dc : 0;
         initial_tablets = std::max(initial_tablets, tablets_in_dc);
     }
     rslogger.debug("Estimated {} initial tablets for table {}.{}", initial_tablets, s.ks_name(), s.cf_name());
     return initial_tablets;
 }
 
-future<tablet_map> network_topology_strategy::allocate_tablets_for_new_table(schema_ptr s, token_metadata_ptr tm, unsigned initial_scale) const {
-    auto tablet_count = get_initial_tablets();
+size_t network_topology_strategy::calculate_min_tablet_count(schema_ptr s, token_metadata_ptr tm, uint64_t target_tablet_size, std::optional<unsigned> initial_scale) const {
+    size_t tablet_count = get_initial_tablets();
+    const auto& tablet_hints = s->tablet_hints();
+    if (tablet_hints.min_tablet_count) {
+        tablet_count = std::max<size_t>(tablet_count, tablet_hints.min_tablet_count.value());
+    }
+    if (tablet_hints.expected_data_size_in_gb) {
+        tablet_count = std::max<size_t>(tablet_count, (tablet_hints.expected_data_size_in_gb.value() << 30) / target_tablet_size);
```

Comment:
we don't have to round up (expected_size / target_size) here since caller will round up to next power of two, right?

Reply:
it's not a bad idea to round up here too

Reply:
yes, it helps for the long run if we ever move away from power-of-two (and forget to update it).

Reply:
Please don't, my PR changes this logic. Rounding up here would hurt the logic: two tables which want 130 tablets and 210 tablets would be treated the same before scaling if we round up early, but we want scaling to affect them proportionally to their impact, and post scaling we can have 130*0.4=52 (64) and 210*0.4=84 (128).

Reply:
the idea here was to only round to the next nearest integer, not to the next power of two. would it still cause a problem? anyway, it's a minor item, so we don't have to do any change here if you prefer.

Reply:
Both std::floor and std::ceil put tablet size within the scheduler's steady state of avg tablet size in the range [target_tablet_size / 2, target_tablet_size * 2]. std::floor prefers fewer and larger tablets, std::ceil prefers more and smaller tablets. I assumed that fewer tablets is preferred.